ziprofs.py

```diff
@@ -162,19 +146,31 @@ def getattr(self, path, fh=None):
 
     def open(self, path, flags):
         zip_path = self.get_zip_path(path)
-        if not zip_path:
-            return os.open(path, flags)
-        return 0
+        if zip_path:
+            with self._lock:
+                fh = self._get_free_zip_fh()
+                zf = self.zip_factory.get(zip_path)
+                self._zip_zfile_fh[fh] = zf
+                self._zip_file_fh[fh] = zf.open(path[len(zip_path) + 1:])
+                return fh
+        else:
+            return os.open(path, flags) << 1
 
     def read(self, path, size, offset, fh):
-        with self.rwlock:
-            zip_path = self.get_zip_path(path)
-            if not zip_path:
-                os.lseek(fh, offset, 0)
-                return os.read(fh, size)
-            zf = self.zip_factory.get(zip_path)
-            subpath = path[len(zip_path)+1:]
-            return zf.read(subpath, size, offset)
+        if fh in self._zip_file_fh:
+            f = self._zip_file_fh[fh]  # should be here (file is first opened, then read)
+            with self._zip_zfile_fh[fh].lock():
+                if not f.seekable():
+                    raise FuseOSError(errno.EBADF)
+
+                if self.log.isEnabledFor(logging.DEBUG):
+                    foffset = f.tell()
+                    self.log.debug(f" file offset: {foffset}, read offset: {offset}, diff: {offset-foffset}")
```

Comment:
Line too long. Is this needed?

Reply:
You mean debug one?
It's not needed, I sed it to.. debug :V
I'll remove this in few minutes.